driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java

```diff
@@ -378,21 +380,46 @@ public <T> T sendAndReceive(final CommandMessage message, final Decoder<T> decod
         if (!Authenticator.shouldAuthenticate(authenticator, this.description)) {
             return sendAndReceiveInternal(message, decoder, sessionContext, requestContext, operationContext);
         }
-        Supplier<T> retryableOperation = () ->
+        Supplier<T> sendAndReceiveInternal = () ->
                 sendAndReceiveInternal(message, decoder, sessionContext, requestContext, operationContext);
         try {
-            return retryableOperation.get();
+            return sendAndReceiveInternal.get();
         } catch (MongoCommandException e) {
             if (triggersReauthentication(e)) {
                 authenticated.set(false);
                 authenticator.reauthenticate(this);
                 authenticated.set(true);
-                return retryableOperation.get();
+                return sendAndReceiveInternal.get();
             }
             throw e;
         }
     }
 
+
+    @Override
+    public <T> void sendAndReceiveAsync(final CommandMessage message, final Decoder<T> decoder, final SessionContext sessionContext,
+            final RequestContext requestContext, final OperationContext operationContext, final SingleResultCallback<T> callback) {
+        notNull("stream is open", stream, callback);
```

Comment:
1. Let's move this line "back" to `sendAndReceiveAsyncInternal`, as otherwise the check stands out when comparing sync and async methods.
1.1. Even better, let's delete the check as the same check is done in `sendMessageAsync`/`sendMessage`, which are eventually called by `sendAndReceiveAsyncInternal`/`sendAndReceiveInternal`.

P.S. I have just realized that all assertions in our async code are incorrect: we should try and pass `AssertionError` to callbacks (provided that this does not lead to completing those twice) instead of propagating them via the stack.